File: backend/app/schemas/target_group.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import SourceFileResponse, TargetGroupPreviewRow, ValidationIssue


class TargetGroupSheetResponse(BaseModel):
    sheet_id: UUID
    source_file_id: UUID | None = None
    sheet_name: str
    sheet_index: int
    sheet_type: str
    row_count: int = 0
    column_names: list[str] = Field(default_factory=list)
    classification_confidence: float | None = None
    notes: str | None = None


class TargetGroupImportSummaryResponse(BaseModel):
    total_uploaded_files: int = 0
    total_rows: int = 0
    parsed_rows: int = 0
    valid_cid_rows: int = 0
    invalid_cid_rows: int = 0
    missing_cid_rows: int = 0
    duplicate_cid_rows: int = 0
    warning_rows: int = 0
    failed_rows: int = 0


class TargetGroupUploadResponse(BaseModel):
    group_id: UUID
    group_name: str
    parse_status: str
    source_file_count: int
    total_rows: int
    import_summary: TargetGroupImportSummaryResponse
    uploaded_files: list[SourceFileResponse] = Field(default_factory=list)
    sheets: list[TargetGroupSheetResponse] = Field(default_factory=list)
    preview_rows: list[TargetGroupPreviewRow] = Field(default_factory=list)
    validation_issues: list[ValidationIssue] = Field(default_factory=list)
    uploaded_at: datetime


class ConfirmImportResponse(BaseModel):
    group_id: UUID
    parse_status: str
    match_status: str


class RunMatchResponse(BaseModel):
    group_id: UUID
    match_status: str
    matched_rows: int
    not_found_rows: int
    ambiguous_rows: int
    needs_review_rows: int


class DiseaseOptionResponse(BaseModel):
    key: str
    label: str
    icd10_code: str | None = None
    raw_name: str | None = None


class MatchSummaryResponse(BaseModel):
    matched: int = 0
    not_found: int = 0
    ambiguous: int = 0
    needs_review: int = 0
    pending: int = 0


class TargetGroupListItemResponse(BaseModel):
    group_id: UUID
    group_name: str
    source_file_name: str
    source_file_type: str
    source_file_count: int = 1
    parse_status: str
    match_status: str
    total_rows: int
    invalid_rows: int
    import_summary: TargetGroupImportSummaryResponse
    match_summary: MatchSummaryResponse
    uploaded_at: datetime


class TargetGroupDetailResponse(BaseModel):
    group_id: UUID
    group_name: str
    source_file_name: str
    source_file_type: str
    source_file_hash: str
    source_set_hash: str | None = None
    source_file_count: int = 1
    parse_status: str
    match_status: str
    total_rows: int
    invalid_rows: int
    import_summary: TargetGroupImportSummaryResponse
    match_summary: MatchSummaryResponse
    uploaded_files: list[SourceFileResponse] = Field(default_factory=list)
    sheets: list[TargetGroupSheetResponse] = Field(default_factory=list)
    preview_rows: list[TargetGroupPreviewRow] = Field(default_factory=list)
    validation_issues: list[ValidationIssue] = Field(default_factory=list)
    uploaded_at: datetime


class UpdateGroupRequest(BaseModel):
    group_name: str = Field(..., min_length=1, max_length=200)


# Returned by POST /{group_id}/add-files — same shape as detail so the
# frontend can refresh the group state from a single response.
AddFilesResponse = TargetGroupDetailResponse


class TargetGroupValidationSummaryResponse(BaseModel):
    group_id: UUID
    total_rows: int
    invalid_rows: int
    missing_cid_rows: int = 0
    duplicate_cid_rows: int = 0
    review_required_rows: int
    validation_issues: list[ValidationIssue] = Field(default_factory=list)

File: frontend/src/types/target-group.ts
export type ValidationIssue = {
  row_id?: string | null;
  row_no: number;
  source_file_id?: string | null;
  source_file_name?: string | null;
  source_row_no?: number | null;
  field: string;
  message: string;
};

export type TargetGroupPreviewRow = {
  row_id?: string | null;
  row_no: number;
  source_file_id?: string | null;
  source_file_name?: string | null;
  source_row_no?: number | null;
  normalized_cid?: string | null;
  parse_status?: string | null;
  values: Record<string, string | number | null>;
};

export type MatchSummary = {
  matched: number;
  not_found: number;
  ambiguous: number;
  needs_review: number;
  pending: number;
};

export type DiseaseOption = {
  key: string;
  label: string;
  icd10_code: string | null;
  raw_name: string | null;
};

export type TargetGroupFile = {
  file_id: string | null;
  file_name: string;
  file_path: string | null;
  file_type: string;
  sha256: string;
  size_bytes: number;
  modified_at: string | null;
  parse_status: string | null;
  row_count: number | null;
  warning_count: number | null;
  error_message: string | null;
  parse_error_summary?: string | null;
};

export type TargetGroupSheet = {
  sheet_id: string;
  source_file_id?: string | null;
  sheet_name: string;
  sheet_index: number;
  sheet_type: string;
  row_count: number;
  column_names: string[];
  classification_confidence?: number | null;
  notes?: string | null;
};

export type TargetGroupImportSummary = {
  total_uploaded_files: number;
  total_rows: number;
  parsed_rows: number;
  valid_cid_rows: number;
  invalid_cid_rows: number;
  missing_cid_rows: number;
  duplicate_cid_rows: number;
  warning_rows: number;
  failed_rows: number;
};

export type TargetGroupUploadResponse = {
  group_id: string;
  group_name: string;
  parse_status: string;
  source_file_count: number;
  total_rows: number;
  import_summary?: TargetGroupImportSummary;
  uploaded_files: TargetGroupFile[];
  sheets: TargetGroupSheet[];
  preview_rows: TargetGroupPreviewRow[];
  validation_issues: ValidationIssue[];
  uploaded_at: string;
};

export type ConfirmImportResponse = {
  group_id: string;
  parse_status: string;
  match_status: string;
};

export type RunMatchResponse = {
  group_id: string;
  match_status: string;
  matched_rows: number;
  not_found_rows: number;
  ambiguous_rows: number;
  needs_review_rows: number;
};

export type TargetGroupListItem = {
  group_id: string;
  group_name: string;
  source_file_name: string;
  source_file_type: string;
  source_file_count: number;
  parse_status: string;
  match_status: string;
  total_rows: number;
  invalid_rows: number;
  import_summary?: TargetGroupImportSummary;
  match_summary: MatchSummary;
  uploaded_at: string;
};

export type TargetGroupDetail = {
  group_id: string;
  group_name: string;
  source_file_name: string;
  source_file_type: string;
  source_file_hash: string;
  source_set_hash: string | null;
  source_file_count: number;
  parse_status: string;
  match_status: string;
  total_rows: number;
  invalid_rows: number;
  import_summary?: TargetGroupImportSummary;
  match_summary: MatchSummary;
  uploaded_files: TargetGroupFile[];
  sheets: TargetGroupSheet[];
  preview_rows: TargetGroupPreviewRow[];
  validation_issues: ValidationIssue[];
  uploaded_at: string;
};

export type TargetGroupValidationSummary = {
  group_id: string;
  total_rows: number;
  invalid_rows: number;
  missing_cid_rows: number;
  duplicate_cid_rows: number;
  review_required_rows: number;
  validation_issues: ValidationIssue[];
};
